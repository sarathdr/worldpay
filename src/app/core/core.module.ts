import { APP_INITIALIZER, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ConfigService } from './services/config.service';
import { PaymentService } from './services/payment.service';
import { CartService } from './services/cart.service';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule } from '@angular/forms';


export const fetchConfig = (appConfig: ConfigService) => {
  return () => appConfig.getPaymentUrls();
};

@NgModule({
  imports: [
    CommonModule,
    FormsModule
  ],
  providers: [
    ConfigService,
    PaymentService,
    HttpClientModule,
    CartService,
    {
      provide: APP_INITIALIZER,
      useFactory: fetchConfig,
      deps: [ConfigService],
      multi: true
    }
  ]
})
export class CoreModule { }
