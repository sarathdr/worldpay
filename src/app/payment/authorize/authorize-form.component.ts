import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Cart } from '../../core/models/cart.interface';
import { NgForm } from '@angular/forms';
import { PaymentInstrument } from '../../core/models/payment.interface';

@Component({
  selector: 'payment-authorize-form',
  template: `
    <form (ngSubmit)="submitForm(authorizeForm)" #authorizeForm="ngForm">
      <mat-form-field>
        <input type="text" matInput placeholder="Card holder name" ngModel name="cardHolderName" required>
        <mat-error>Enter card holder name</mat-error>
      </mat-form-field>
      <mat-form-field>
        <input matInput placeholder="Card number" ngModel name="cardNumber" type="number" required>
        <mat-error>Enter a valid card number</mat-error>
      </mat-form-field>
      <mat-label>Expiry date</mat-label>
      <div class="row" ngModelGroup="cardExpiryDate">
        <mat-form-field class="month">
          <mat-label>MM</mat-label>
          <mat-select name="month" ngModel required #month="ngModel">
            <mat-option *ngFor="let item of months" [value]="item">
              {{ item }}
            </mat-option>
          </mat-select>
          <mat-error>Select month</mat-error>
        </mat-form-field>
        <mat-form-field>
          <input
              type="text"
              name="year"
              placeholder="YYYY"
              ngModel
              matInput
              required
              pattern="\\d*"
              maxlength="4"
              minlength="4"
              expiryDate
              [month]="month.value"
              #year="ngModel"
          />
          <mat-error>Enter a valid date</mat-error>
        </mat-form-field>
      </div>
      <div class="row">
        <button mat-raised-button type="button" routerLink="/home">Cancel</button>
        <button mat-raised-button color="primary" type="submit">Next</button>
      </div>
    </form>
  `,
  styles: [
      `
          form {
              display: flex;
              flex-direction: column;
              justify-content: space-between;
              width: 400px;
              height: 285px;
              margin: 0 auto;
          }

          .row {
              display: flex;
              flex-direction: row;
              justify-content: space-between;
          }

          .month {
              width: 100px;
          }
    `
  ]
})
export class AuthorizeFormComponent {
  readonly months = [1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12];

  @Output() onSubmit = new EventEmitter<PaymentInstrument>();

  submitForm(authorizeForm: NgForm) {
    if (authorizeForm.valid) {
      console.log('form is valid');
      this.onSubmit.emit(authorizeForm.value);
    }
  }
}
