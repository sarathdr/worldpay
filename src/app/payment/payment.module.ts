import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AuthorizeComponent } from './authorize/authorize.component';
import { AuthorizeContainer } from './authorize/authorize.container';
import { ConfirmComponent } from './confirm/confirm.component';
import { ConfirmContainer } from './confirm/confirm.container';
import { RefundComponent } from './refund/refund.component';
import { RefundContainer } from './refund/refund.container';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { MatButtonModule, MatFormFieldModule, MatInputModule, MatSelectModule } from '@angular/material';
import { AuthorizeFormComponent } from './authorize/authorize-form.component';
import { ConfirmPaymentGuard } from './guards/confirm-payment.guard';
import { ExpiryDateModule } from '../core/directives/expiry-date.module';
import { RefundPaymentGuard } from './guards/refund-payment.guard';
import { RefundSuccessComponent } from './refund/refund-success.component';

@NgModule({
  declarations: [
    AuthorizeComponent,
    AuthorizeFormComponent,
    AuthorizeContainer,
    ConfirmComponent,
    ConfirmContainer,
    RefundComponent,
    RefundSuccessComponent,
    RefundContainer
  ],
  imports: [
    CommonModule,
    FormsModule,
    RouterModule.forChild([
      {
        path: '',
        component: AuthorizeContainer
      },
      {
        path: 'confirm',
        component: ConfirmContainer,
        canActivate:[ConfirmPaymentGuard],
        canDeactivate:[ConfirmPaymentGuard]
      },
      {
        path: 'refund',
        component: RefundContainer,
        canActivate:[RefundPaymentGuard],
        canDeactivate:[RefundPaymentGuard]
      }
    ]),
    MatFormFieldModule,
    MatInputModule,
    MatSelectModule,
    ReactiveFormsModule,
    MatButtonModule,
    ExpiryDateModule
  ],
  providers:[
    ConfirmPaymentGuard,
    RefundPaymentGuard
  ]
})
export class PaymentModule { }
