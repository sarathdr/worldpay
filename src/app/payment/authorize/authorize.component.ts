import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Cart } from '../../core/models/cart.interface';
import { Instruction, InstrumentType, PaymentInstrument } from '../../core/models/payment.interface';

@Component({
  selector: 'payment-authorize',
  template: `
    <payment-authorize-form
        (onSubmit)="onSubmit($event)">
    </payment-authorize-form>
  `
})
export class AuthorizeComponent {

  @Input() cart: Cart;
  @Output() authorizePayment = new EventEmitter<Instruction>();

  onSubmit(paymentInstrument: PaymentInstrument) {
    const instruction: Instruction =  {
      narrative: {
        line1: 'trading name'
      },
    value: {
        amount: this.cart.total,
        currency: this.cart.currency
    },
    paymentInstrument: {
      ...paymentInstrument,
     type: InstrumentType.CARD_PLAIN
    }
  };
    this.authorizePayment.emit(instruction);
  }
}
