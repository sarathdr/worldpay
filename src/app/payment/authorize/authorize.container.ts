import { Component, OnDestroy } from '@angular/core';
import { Observable, Subscription } from 'rxjs';
import { Cart } from '../../core/models/cart.interface';
import { CartService } from '../../core/services/cart.service';
import { PaymentService } from '../../core/services/payment.service';
import { AuthorizeRequest, Instruction } from '../../core/models/payment.interface';
import { v4 as uuid } from 'uuid';
import { Router } from '@angular/router';
import { take } from 'rxjs/operators';

@Component({
  selector: 'payment-authorize-page',
  template: `
    <payment-authorize
        [cart]="cart$ | async" (authorizePayment)="authorize($event)">
    </payment-authorize>
  `
})
export class AuthorizeContainer {

  cart$: Observable<Cart>;

  constructor(private cartService: CartService,
              private paymentService: PaymentService,
              private router: Router) {
    this.cart$ = cartService.getCart();
  }

  authorize(instruction: Instruction) {
    const body: AuthorizeRequest = {
      transactionReference: uuid(),
      merchant: {
        entity: 'default'
      },
      instruction
    };

    this.paymentService.authorizePayment(body)
      .pipe(take(1))
      .subscribe(() => this.router.navigate(['payment', 'confirm']));
  }
}
