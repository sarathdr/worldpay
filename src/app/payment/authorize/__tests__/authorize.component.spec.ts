import { TestBed } from '@angular/core/testing';
import { Component, EventEmitter, Output } from '@angular/core';
import { MatButtonModule, MatListModule, MatTableModule } from '@angular/material';
import { RouterTestingModule } from '@angular/router/testing';
import { AuthorizeComponent } from '../authorize.component';
import { Currency, InstrumentType, PaymentInstrument } from '../../../core/models/payment.interface';
import { By } from '@angular/platform-browser';

describe('AuthorizeComponent', () => {

  let fixture;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        MatListModule,
        MatButtonModule,
        MatTableModule
      ],
      declarations: [
        AuthorizeComponent,
        MockAuthorizeFormComponent,
        TestAuthorizeComponent
      ],
    });

    fixture = TestBed.createComponent(TestAuthorizeComponent);
  });

  it('should render the component correctly', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should handle submit form', () => {
    fixture.detectChanges();

    const formComponent = fixture.debugElement.query(By.directive(MockAuthorizeFormComponent)).componentInstance;
    formComponent.onSubmit.emit({
      type: InstrumentType.CARD_PLAIN,
      cardHolderName: 'testing',
      cardNumber: 4235234534,
      cardExpiryDate: {
        month: 5,
        year: 2020
      }
    });

    expect(fixture.componentInstance.authorizePayment.mock.calls).toMatchSnapshot();
  });

  @Component({
    selector: 'payment-authorize-form',
    template: `
      <div>Authorize form</div>
    `
  })
  class MockAuthorizeFormComponent {
    @Output() onSubmit = new EventEmitter<PaymentInstrument>();
  }

  @Component({
    selector: 'test-authorize-component',
    template: `
      <payment-authorize
          [cart]="cart"
          (authorizePayment)="authorizePayment($event)">
      </payment-authorize>`
  })
  class TestAuthorizeComponent {
    cart = {
      products: [{
        name: 'Pen',
        price: 10
      },
        {
          name: 'Pencil',
          price: 12
        }],
      total: 22,
      currency: Currency.GBP
    };

    authorizePayment = jest.fn();
  }

});
