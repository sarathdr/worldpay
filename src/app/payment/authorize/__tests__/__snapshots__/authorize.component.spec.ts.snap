// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorizeComponent should handle submit form 1`] = `
Array [
  Array [
    Object {
      "narrative": Object {
        "line1": "trading name",
      },
      "paymentInstrument": Object {
        "cardExpiryDate": Object {
          "month": 5,
          "year": 2020,
        },
        "cardHolderName": "testing",
        "cardNumber": 4235234534,
        "type": "card/plain",
      },
      "value": Object {
        "amount": 22,
        "currency": "GBP",
      },
    },
  ],
]
`;

exports[`AuthorizeComponent should render the component correctly 1`] = `
<test-authorize-component
  authorizePayment={[Function Function]}
  cart={[Function Object]}
>
  <payment-authorize
    ng-reflect-cart="[object Object]"
  >
    <payment-authorize-form>
      <div>
        Authorize form
      </div>
    </payment-authorize-form>
  </payment-authorize>
</test-authorize-component>
`;
